hawkular-command-gateway-war/pom.xml

```diff
@@ -72,11 +72,6 @@
     </dependency>
 
     <dependency>
-      <groupId>org.hawkular.inventory</groupId>
-      <artifactId>hawkular-inventory-json-helper</artifactId>
-    </dependency>
```

Comment:
I tend to think the client will NOT know that. A client should only have to know of a resource it wants to manipulate - it should never care about the underlying management infrastructure (in other words, it shouldn't have to know about what management agent or feed is managing a resource - we should just have to require it to tell us what resource it wants to operate on).

Also, a client may be sending a message to a GROUP of feeds - as would be the case when, in the future, we have a JSON command like "ExecuteOperationOnGroup" where the command will have an attribute "groupPath" (groupPath could be just like an inventory resourcePath only it points to a path that refers to a group of resources defined in inventory). The client won't know all the feeds that that manage all the resources in that group. Even if it did, this mechanism doesn't allow for multiple receipients. But again, we should stress that the client should never care about what agents or feeds are managing resources - we should just ask them to provide us with the RESOURCE identifiers.

This is why I liked the idea of the command implementations on the server being the ones to determine where to forward the messages. Because some commands a client will tell us a single resourcePath (which are then parsed via the inventory API to determine what feed is managing that resource), others may provide a group path (which could then be used to query the inventory to determine which resources are currently in that group, and then from that which feed(s) manage those resources keeping in mind one feed many manage more than one of these resources), others may be an array of resourcePaths. The point is, don't make this solution so generic that we lose the ability to customize our command implementations in the future. As it stands now, this only supports messages to one recipient, and it requires the client to know what that receiptient's feed ID is.

So what I would like to see rather than have one "general" class to handle all the forwarding, perhaps have a superclass that all command execution classes extend - this superclass would provide utility methods like "forwardToFeed" or "forwardToBus" but have (again, as an example) an abstract method called "obtainRecipientIDs(BasicMessage command)" that must be implemented by the command - the command would know "I am the ExecuteOperationByGroup" command, so I know I should look for the "groupPath" attribute, and from there I need to ask inventory what feeds manage the resources in that group".

Or, as an alternative, perhaps we have some basic command design patterns that we look for. Maybe this one uber class would look for one of several types of addressing JSON elements. For example, maybe it says "is there a 'resourcePath' attribute? If so, use that to find the one feed ID as the recipient. If there is no resourcePath, is there a 'resourcePaths' attribute (notice the "s") - if so, use that to get the multiple feed IDs that manage those multiple resources. In the future we could add "does the JSON command have a "groupPath" attribute - if so, use that and determine the feeds that manage that group of resources"

In this way, you don't need to have these special "receipientId" things in the special AuthMessage. We just have a design pattern where each command must have one of the several different attributes that are supported (like "resourcePath" or "resourcePaths" or "groupPath"). Note that all of commands already have this, except for EchoCommand - which isn't targeting any feed today but should be (right now EchoCommand just echos from the server, it doesn't forward it to the agent so it isn't the agent that is getting pings via the Echo its just the server. But the Echo stuff is just for testing anyway. So we can do whatever for that.)

Hopefully this makes sense :)